src/ConnectionPool.jl

```diff
@@ -70,17 +69,19 @@ mutable struct Connection{T <: IO}
     peerport::UInt16 # debug only
     localport::UInt16 # debug only
     io::T
+    clientconnection::Bool
     buffer::IOBuffer
     sequence::Int
     writecount::Int
-    writebusy::Bool
-    writedone::Condition
+    writelock::Channel{Bool}
```

Comment:
This probably either has the wrong name or wrong type. Seems like you meant to use a `RecursiveLock` here?

Reply:
I'm not familiar w/ `RecursiveLock`; it doesn't seem to be defined in Base? I agree that the "writelock" here is a bit awkward; the idea is that the `Connection` object only wants to allow one Transaction at a time to have a read lock or write lock. So Transactions want to acquire the lock in `startwrite`/`startread`, and wait if it's already locked by another transaction. A `Channel` does the `take!`-ing part of what I want, but we don't really care about what we're getting from the Channel.

Could you point me to wherever RecursiveLock is defined?

Reply:
`Threads.ReentrantLock` perhaps?

Reply:
The issue with ReentrantLock is that the locking/unlocking may occur in different tasks (though wouldn't ever occur in different threads); this is because the lock tends to happen in the "main" thread doing the toplevel request, whereas the unlock (may) occur in the `@async writebody` task. I tried this before and ReentrantLock complained about the wrong task unlocking it.

I found https://www.justsoftwaresolutions.co.uk/threading/locks-mutexes-semaphores.html, which seems like a pretty good walk-through of basic threading primitives. From that, it feels like what we want here is a Semaphore, since we want the locking and signaling functionality, but are not as concerned with "who" does the acquire/release, since we're managing that logic ourselves through our `Transaction` structure.

Reply:
Ok, after some more mulling over, I switched our lock structure here back to a `Condition`. Correct me if I'm wrong, but after studying the `Channel` implementation in Base, a `Condition` is like a lock that can also signal to waiters, which is basically what we want here. So we have `writelock::Condition` and `writebusy::Threads.Atomic{Bool}`, so query-ers can check if a Connection is busy by checking `conn.writebusy[]`, and we manage the setting/unsetting of `writebusy` in `startwrite`/`closewrite` exclusively. We also currently have the `writelock` locked when updating `writebusy` and `writecount`, even though they're atomics anyway, but just to more document the data we're "locking" updates on.